### PR TITLE
[Dicom-cast] Continue processing of change feed after intransient errors

### PR DESCRIPTION
**Pull request by abdullah248**
## Description
Intransient errors (non-retriable errors) should all now get logged into table storage and then continue with processing of the next change feed entry. Also if partial validation is enabled then in the case of non required meta data from change feed entry being invalid, then store the error to table storage and still process the change feed entry and send to fhir.

I updated the table storage to have 1 table for dicom validation errors in fields that are not required and then 1 table for all errors that cause intransient errors (including fhir errors, dicom validation errors for fields that are required, and dicom validation errors for non required fields when partial validation is disabled)

This also has the changes for the pr #492. I closed that pr with out merging as all the changes are here and I updated the comments from that pr here. 

## Related issues
[AB#78074](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/78074)
[AB#78068](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/78068)

## Testing
Added unit tests where possible and manually tested to check the entire flow

### CODE REVIEW

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/ExceptionStorage/IExceptionStore.cs [hunk 1]
```diff
@@ -23,6 +24,6 @@ public interface IExceptionStore
         /// <param name="exceptionToStore">The exception that was thrown and needs to be stored</param>
         /// <param name="errorType">The type of error thrown</param>
         /// <param name="cancellationToken">Cancellation token.</param>
-        void StoreException(string studyUid, string seriesUid, string instanceUid, long changeFeedSequence, Exception exceptionToStore, ErrorType errorType, CancellationToken cancellationToken = default);
+        public Task StoreException(string studyUid, string seriesUid, string instanceUid, long changeFeedSequence, Exception exceptionToStore, ErrorType errorType, CancellationToken cancellationToken = default);
```
Comment (smithago) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/ExceptionStorage/IExceptionStore.cs:27]: StoreExceptionAync?

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyInstancePropertySynchronizer.cs [hunk 2]
```diff
@@ -26,14 +44,54 @@ public void Synchronize(FhirTransactionContext context, ImagingStudy.InstanceCom
...
+            catch (Exception ex)
+            {
+                if (_dicomValidationConfiguration.PartialValidation && !required)
+                {
+                    DicomDataset dataset = context.ChangeFeedEntry.Metadata;
+                    string studyUID = dataset.GetSingleValue<string>(DicomTag.StudyInstanceUID);
+                    string seriesUID = dataset.GetSingleValue<string>(DicomTag.SeriesInstanceUID);
+                    string instanceUID = dataset.GetSingleValue<string>(DicomTag.SOPInstanceUID);
+
+                    _exceptionStore.StoreException(
```
Comment (smithago) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyInstancePropertySynchronizer.cs:66]: Will we not await on this?

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyInstancePropertySynchronizer.cs [hunk 3]
```diff
@@ -26,14 +44,54 @@ public void Synchronize(FhirTransactionContext context, ImagingStudy.InstanceCom
...
+            try
+            {
+                synchronizeAction(instance, context);
+            }
+            catch (Exception ex)
+            {
+                if (_dicomValidationConfiguration.PartialValidation && !required)
+                {
+                    DicomDataset dataset = context.ChangeFeedEntry.Metadata;
+                    string studyUID = dataset.GetSingleValue<string>(DicomTag.StudyInstanceUID);
```
Comment (smithago) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyInstancePropertySynchronizer.cs:62]: nit: studyUid

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyInstancePropertySynchronizer.cs [hunk 4]
```diff
@@ -26,14 +44,54 @@ public void Synchronize(FhirTransactionContext context, ImagingStudy.InstanceCom
...
+                    string seriesUID = dataset.GetSingleValue<string>(DicomTag.SeriesInstanceUID);
+                    string instanceUID = dataset.GetSingleValue<string>(DicomTag.SOPInstanceUID);
+
+                    _exceptionStore.StoreException(
+                        studyUID,
+                        seriesUID,
+                        instanceUID,
+                        context.ChangeFeedEntry.Sequence,
+                        ex,
+                        ErrorType.DicomValidationError,
```
Comment (smithago) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyInstancePropertySynchronizer.cs:72]: How are we categorizing this has validationError? We are catching all exceptions here

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyPropertySynchronizer.cs [hunk 5]
```diff
@@ -30,17 +48,47 @@ public void Synchronize(FhirTransactionContext context, ImagingStudy imagingStud
...
+        private void SynchronizePropertiesAsync(ImagingStudy imagingStudy, FhirTransactionContext context, bool required, Action<ImagingStudy, FhirTransactionContext> synchronizeAction, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                synchronizeAction(imagingStudy, context);
+            }
+            catch (Exception ex)
+            {
+                if (_dicomValidationConfiguration.PartialValidation && !required)
+                {
```
Comment (smithago) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyPropertySynchronizer.cs:67]: same here

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Patient/IPatientPropertySynchronizer.cs [hunk 6]
```diff
@@ -20,5 +20,10 @@ public interface IPatientPropertySynchronizer
         /// <param name="patient">The <see cref="Patient"/> resource.</param>
         /// <param name="isNewPatient">Flag to determine whether or not the patient being synchronized is new.</param>
         void Synchronize(DicomDataset dataset, Patient patient, bool isNewPatient);
+
+        /// <summary>
+        /// Returns if the property is a required property for the <see cref="Patient"/> resource
+        /// </summary>
+        bool IsRequired();
```
Comment (smithago) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Patient/IPatientPropertySynchronizer.cs:27]: Can you add a little details on how this property is used? I see all the properties are marked false

Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Patient/IPatientPropertySynchronizer.cs:27]: I did it based on the cardinality from this link: https://www.hl7.org/fhir/patient.html.  

I used it in the patientsynchronizer where we loop through all the properties. In the case that a property is not required then we we just store to table storage and try to sync the rest of the properties.

Looking further into the spec thou, it actually seems like all of the elements for patient resource are not required so I can probably remove this.

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.json [hunk 7]
```diff
@@ -37,6 +37,9 @@
         "PollInterval": "00:01:00",
         "PollIntervalDuringCatchup": "00:00:00"
     },
+    "DicomValidation": {
```
Comment (smithago) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.json:40]: Should this property be in DicomCastWroker and maybe have a name like "IgnoreSyncOfInvalidTagValue"?

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.json [hunk 8]
```diff
@@ -37,6 +37,9 @@
...
+        "PartialValidation" : false
```
Comment (smithago) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.json:41]: Why is it false here and true in development json

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyInstancePropertySynchronizer.cs (hunk 4 above)
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyInstancePropertySynchronizer.cs:72]: So my thought here is that each time we are executing the Action we are trying to sync a specific property, and if while doing that on a field that is not required we can just store the error and continue to the next one. 

Technically these all aren't validation errors but if it is not a required field can't we continue processing regardless? Or do we want to handle those errors differently?

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Patient/PatientSynchronizer.cs [hunk 9]
```diff
@@ -16,23 +22,63 @@ namespace Microsoft.Health.DicomCast.Core.Features.Worker.FhirTransaction
...
+            DicomDataset dataset = context.ChangeFeedEntry.Metadata;
+
             foreach (IPatientPropertySynchronizer patientPropertySynchronizer in _patientPropertySynchronizers)
             {
-                patientPropertySynchronizer.Synchronize(dataset, patient, isNewPatient);
+                try
+                {
+                    patientPropertySynchronizer.Synchronize(dataset, patient, isNewPatient);
+                }
+                catch (Exception ex)
```
Comment (smithago) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Patient/PatientSynchronizer.cs:60]: Is this covered with unit tests?

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.json (hunk 8 above)
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.json:41]: I removed it from development json and left it as false here for the default.

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyInstancePropertySynchronizer.cs (hunk 2 above)
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyInstancePropertySynchronizer.cs]: updated, thanks

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.json (hunk 7 above)
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.json:40]: I could move it there, but would we want it to be a separate configuration class still? Currently the DicomCastWorkerConfiguration is only used in DicomCastWorker, where as the DicomValidationConfiguration is used in multiple other locations, such as the patient and imaging property synchronizers. I don't think we would want to pass the whole DicomCastWorkerConfiguration to those places.

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Patient/PatientSynchronizer.cs (hunk 9 above)
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Patient/PatientSynchronizer.cs:56]: I added unit tests to PatientSynchronizerTests for this method and to ImagingStudyPipelineHelperTests for some other changes, looking for other places to add testing coverage as well.

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.json (hunk 7 above)
Comment (smithago) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.json:40]: Sure. The name is still throwing me off, the names should represent customer scenario, not what the implementation does. So maybe
"DicomCast": {
    "Features": {
        "IgnoreSyncOfInvalidTagValue": false
    }
}

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Patient/IPatientPropertySynchronizer.cs (hunk 6 above)
Comment (smithago) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Patient/IPatientPropertySynchronizer.cs:27]: Lets remove it, only required is patientId I think

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Patient/PatientSynchronizer.cs (hunk 9 above)
Comment (smithago) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Patient/PatientSynchronizer.cs:56]: Sounds good. Is there a specific expcetion type you catch here. Lets say you catch a OutOfMemory exception here, you don't want to handle it. You will let it bubble up.

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.json (hunk 7 above)
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.json]: Updated to as suggested

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Patient/IPatientPropertySynchronizer.cs (hunk 6 above)
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Patient/IPatientPropertySynchronizer.cs]: removed

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyInstancePropertySynchronizer.cs (hunk 4 above)
Comment (abdullah248) [reply converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/ImagingStudy/ImagingStudyInstancePropertySynchronizer.cs]: Updated to catch more specific validation error.

**Review by smithago — APPROVED** [review]

**Pull request merged**